src/ripple/nodestore/impl/DatabaseShardImp.h

```diff
@@ -178,17 +178,56 @@ class DatabaseShardImp : public DatabaseShard
             final,      // Immutable, complete and validated
             acquire,    // Being acquired
             import,     // Being imported
+            postImport, // About to be finalized
             finalize    // Being finalized
         };
 
         ShardInfo() = default;
         ShardInfo(std::shared_ptr<Shard> shard_, State state_)
             : shard(std::move(shard_))
             , state(state_)
+            , lastLedgerHashAttempts(0)
         {}
 
         std::shared_ptr<Shard> shard;
         State state {State::none};
+
+        /////////////////////////////////////////////////
+        // The following members are used only for shards
+        // that are being imported.
+        // ----------------------------------------------
+        /////////////////////////////////////////////////
+
+        // Used to limit the number of times we attempt
+        // to retrieve a shard's last ledger hash, when
+        // the hash should have been found. See
+        // scheduleFinalizeShard(). Once this limit has
+        // been reached, the shard has failed and is
+        // removed.
+        bool
+        attemptHashConfirmation()
+        {
+            if (lastLedgerHashAttempts + 1 <= maxLastLedgerHashAttempts)
+            {
+                ++lastLedgerHashAttempts;
+                return true;
+            }
+
+            return false;
+        }
+
+        // This variable is used during the validation
+        // of imported shards and must match the
+        // imported shard's last ledger hash.
+        uint256 lastLedgerHash;
+
+        // The number of times we've attempted to
+        // confirm this shard's last ledger hash.
+        uint16_t lastLedgerHashAttempts;
+
+        // The upper limit on attempts to confirm
+        // the shard's last ledger hash.
+        static const uint8_t maxLastLedgerHashAttempts = 5;
```

Comment:
This can can be `constexpr`. eg. `static constexpr uint8_t maxLastLedgerHashAttempts = 5;`

Reply:
Done